### PR TITLE
Alert of package development guidelines [av skip]

### PR DESCRIPTION
**Pull request by milktrader**

### CODE REVIEW

**Comment by tkelman** [comment]
lgtm, anything else you want to add to?

**Comment by milktrader** [comment]
Probably polish up the naming conventions section at some point, but this area has some controversy so it would probably deserve a gestation period.

**Comment by tkelman** [comment]
did you mean to close this?

**Comment by milktrader** [comment]
Oops, I thought you merged it. I was looking at the wrong merge icon

**Comment by milktrader** [comment]
Yeah, best to get some community feedback for documentation changes that involve more than just typos :smirk:

**Comment by ahojukka5** [comment]
I personally think that every package name should always be accepted unless it is something very disturbing and purposely aggressive (including swearwords etc.).

We might be actually talking about two different things here and mixing them unsuccesfully together. For branding purposes we want to talk about "JuliaFEM" because "FEM" would be way too general software name in scope of ALL software projects around world and probably conflicting. THAT is the reason why we want to use prefix "Julia" in the package name. When we are talking about "JuliaFEM" people immediately understands that it's a finite element solver programmed using Julia. 

If it is really wanted that there is no words like "Julia" in package name, how about doing "package aliases", i.e. user can type
> using JuliaFEM
or
> using FEM
to use package when there is a word "Julia" in package?

I think in Python it's totally ok to use different package name for branding purposes and include it using different name. Name of the actual imported package is defined in setup.py and several packages can be in one repository. It should also be possible in Julia, i.e. user can

> Pkg.add("JuliaFEM")
> using FEMSolvers
> using FEMPostProcess

so that user doesn't necessarily have to compile the whole package when e.g. post processing is not needed. I think this would also speed up code execution time in larger projects because user can import just what is needed at that time. I have developed a lot of Python packages and noticed that if all modules are imported in __init__.py, it slows down starting the package. For that reason I usually define several packages or modules, e.g. matplotlib is imported only when there is a need to plot something because it takes some time to import matplotlib.

I tried to cp JuliaFEM.jl FEM.jl && using FEM but it's not working.

**Comment by johnmyleswhite** [comment]
> I personally think that every package name should always be accepted unless it is something very disturbing and purposely aggressive (including swearwords etc.).

-100 for this. We've had a long cultural tradition of doing this and it's led to Julia projects having much more clear names than other languages.

**Comment by tkelman** [comment]
Renaming a package does require source changes everywhere you refer to the module name.

I'm pretty sure there's a different issue open for doing something like `using FiniteElement as FEM`, but there's much less need to refer to the module name explicitly in Julia unless you have exported name collisions with other packages. Hence there's a strong desire to avoid the package namespace being too filled with impenetrable acronyms. Sure FEM is a moderately well-known field, but it's not entirely unambiguous and it's better for discoverability to have a clear name.

Package names all end in `.jl` in terms of their github repositories, and the name you would refer to them outside of Julia. Inside of Julia you leave off the `.jl` part of the package name to get the module name, and the context is clear. We have only a handful of packages with Julia in the name - JuliaParser which is a special case because what it does is parse Julia code, then IJulia, SimJulia, and JuliaWebRepl which have been around for some time.

**Comment by tkelman** [comment]
There has been some discussion (at JuliaCon and before) about better supporting a "free-for-all" second tier alternative to METADATA that undergoes little to no scrutiny. In addition to needing to design how that system would work and implement support for it in the package manager, the process of "graduating" from the free-for-all to the curated subset would need to be documented as clearly as we can. Renaming a package to match the higher standards the community expects of the curated subset would occasionally have to happen.

It's not that big a deal - and in the particular case that spurred this, JuliaFEM is a perfect organization name, there's no strong requirement that the headline package in an organization have the same name as the organization itself.

**Comment by ahojukka5** [comment]
In Python there's 62000 packages in pypi.org and believe or not, it's working well without package naming control. Do you really want to have this discussion 62000 times in future? In npm there's 160000 packages. Again do you really want to control naming of 160000 packages?

Like said yesterday, it's a very frustrating situation to find out that after you end up to some name and like it someone is starting to demand that it needs to be changed in order to get package registered. Some other would say that "keep your package index" and start doing Pkg.clone instead. Or start to develop own package index without naming control. In my view it's just not fitting to the spirit of open source to control the project names of other people for unnecessary reasons.

**Comment by johnmyleswhite** [comment]
But Julia isn't Python. We're a young language that needs to develop a coherent ecosystem. We don't yet have any need for the long tail of small packages that exists in Node. A little bikeshedding helps to ensure that a large communal resource (the package namespace) is used well.

> Like said yesterday, it's a very frustrating situation to find out that after you end up to some name and like it someone is starting to demand that it needs to be changed in order to get package registered.

I don't think the situation is objectively frustrating. It can be, but it generally doesn't upset most people. I certainly don't think it violates the spirit of open source.

**Comment by tkelman** [comment]
And on the other extreme, we heard a few days ago about how CRAN denies packages because of a missing period at the end of a sentence in the documentation. We're working out our own system that's somewhere in between.

No one is demanding anything - if you had looked through some of the community discussions that happen often in METADATA, sent a message to the mailing list saying "hey we're working on a package for finite element analysis, what do people think," you would have gotten a sense that this bikeshedding process happens often. It's harmless.

**Comment by milktrader** [comment]
`Pkg.add(;repos="https://github.com/JuliaLang/METADATA.jl")` for the curated version and a separate url for a free-for-all maybe?

Yeah, CRAN is quite a bit more autocratic than METADATA and because of this many good packages end up on different repositories such as R-Forge. 

This probably deserves its own issue though.

**Comment by milktrader** [comment]
https://github.com/JuliaLang/julia/issues/11914

**Comment by mauro3** [comment]
Maybe it should be explicitly stated down in the package naming section that there is no guarantee that ones preferred name will be included in METADATA.jl and that it is up for discussion before inclusion.

A bit off-topic, should we make it a stated requirement that packages in METADATA need to be open source?

**Comment by milktrader** [comment]
Yeah, good point about the no guarantee about names. This should take out the element of surprise. At least a heads up that one can expect some bike-shedding and what to expect. I can do that now.

**Comment by tkelman** [comment]
> should we make it a stated requirement that packages in METADATA need to be open source?

I think enforcing an OSI-approved license in the LICENSE file would be reasonable.

**Comment by milktrader** [comment]
https://github.com/JuliaLang/julia/pull/11916

**Comment by StefanKarpinski** [comment]
> I think enforcing an OSI-approved license in the LICENSE file would be reasonable.

+:100:

File: doc/manual/packages.rst [hunk 1]
```diff
@@ -11,6 +11,10 @@ It can also install external libraries using your operating system's standard sy
 The list of registered Julia packages can be found at `<http://pkg.julialang.org>`_.
 All package manager commands are found in the :mod:`Pkg <Base.Pkg>` module, included in Julia's :mod:`Base` install.
 
+First we'll go over the mechanics of the ``Pkg`` family of commands and then we'll provide some guidance on how
+to get your package registered. Be sure to read the section below on package naming conventions, tagging versions
+and the importance of a ``REQUIRE`` file for when you're ready to add you code to the curated METADATA repository.
```
Comment (tkelman) [inline doc/manual/packages.rst]: `to add your code`

**Comment by tkelman** [comment]
Thanks @milktrader, I think we'll want to make the same change on master too.

**Comment by milktrader** [comment]
Is that a separate PR?

**Comment by tkelman** [comment]
Yeah, or someone with commit access can cherry-pick it if it applies cleanly

**Comment by milktrader** [comment]
Hmm, that doesn't seem very DRY.

**Pull request merged**